substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_java_nio_Bits.java

```diff
@@ -53,3 +55,23 @@ final class Target_java_nio_Bits {
 
     // Checkstyle: resume
 }
+
+/**
+ * {@code java.nio.Bits} caches the max. direct memory size in the field {@code MAX_MEMORY}. We
+ * disable this cache and always call {@link DirectMemoryAccessors#getDirectMemory()} instead, which
+ * uses our own cache. Otherwise, it could happen that {@code MAX_MEMORY} caches a temporary value
+ * that is used during early VM startup, before {@link PhysicalMemory} is fully initialized.
+ */
+final class MaxMemoryAccessor {
+    // Checkstyle: stop
+
+    static long getMAX_MEMORY() {
+        return DirectMemoryAccessors.getDirectMemory();
+    }
+
+    static void setMAX_MEMORY(long value) {
+        /* Nothing to do. */
+    }
+
+    // Checkstyle: resume
+}
```

Comment:
Instead of going through the accessor an alternative could be to leaverage the init hooks in JDK code and actually set the init level to `1` once we know the real `MaxDirectMemorySize` (base on the heap size). For example here: https://github.com/oracle/graal/pull/7553/files#diff-db72af86eee732c8a7e3e53319b32fdd92ee0ec00735f413a02858c9dc0aff4dR126. That would mimic what HotSpot itself is doing. FWIW, that's what my original code was intending to do.

Both approaches work.

Reply:
The problem is that Native Image is already multi-threaded when this code is executed. Modifying the init level may have unexpected effects on other threads.

Reply:
OK, thanks.